crates/ruff_cli/src/cache.rs

```diff
@@ -1,171 +1,240 @@
-use std::cell::RefCell;
-use std::fs;
+use std::collections::HashMap;
+use std::fs::{self, File};
 use std::hash::Hasher;
-use std::io::Write;
-#[cfg(unix)]
-use std::os::unix::fs::PermissionsExt;
-use std::path::Path;
+use std::io::{self, BufReader, BufWriter, Write};
+use std::path::{Path, PathBuf};
+use std::sync::Mutex;
+use std::time::SystemTime;
 
-use anyhow::Result;
-use filetime::FileTime;
-use log::error;
-use path_absolutize::Absolutize;
-use ruff_text_size::{TextRange, TextSize};
-use serde::ser::{SerializeSeq, SerializeStruct};
-use serde::{Deserialize, Serialize, Serializer};
+use anyhow::{anyhow, Context, Result};
+use serde::{Deserialize, Serialize};
 
 use ruff::message::Message;
-use ruff::settings::{AllSettings, Settings};
+use ruff::settings::Settings;
 use ruff_cache::{CacheKey, CacheKeyHasher};
 use ruff_diagnostics::{DiagnosticKind, Fix};
 use ruff_python_ast::imports::ImportMap;
 use ruff_python_ast::source_code::SourceFileBuilder;
+use ruff_text_size::{TextRange, TextSize};
 
-const CARGO_PKG_VERSION: &str = env!("CARGO_PKG_VERSION");
-
-/// Vec storing all source files. The tuple is (filename, source code).
-type Files<'a> = Vec<(&'a str, &'a str)>;
-type FilesBuf = Vec<(String, String)>;
-
-struct CheckResultRef<'a> {
-    imports: &'a ImportMap,
-    messages: &'a [Message],
-}
-
-impl Serialize for CheckResultRef<'_> {
-    fn serialize<S>(&self, serializer: S) -> std::result::Result<S::Ok, S::Error>
-    where
-        S: Serializer,
-    {
-        let mut s = serializer.serialize_struct("CheckResultRef", 3)?;
-
-        s.serialize_field("imports", &self.imports)?;
-
-        let serialize_messages = SerializeMessages {
-            messages: self.messages,
-            files: RefCell::default(),
+use crate::diagnostics::Diagnostics;
+
+/// On disk representation of a cache of a package.
+#[derive(Deserialize, Debug, Serialize)]
+pub(crate) struct PackageCache {
+    /// Location of the cache.
+    ///
+    /// Not stored on disk, just used as a storage location.
+    #[serde(skip)]
+    path: PathBuf,
+    /// Path to the root of the package.
+    ///
+    /// Usually this is a directory, but it can also be a single file in case of
+    /// single file "packages", e.g. scripts.
+    package_root: PathBuf,
+    /// Mapping of source file path to it's cached data.
+    // TODO: look into concurrent hashmap or similar instead of a mutex.
+    files: Mutex<HashMap<RelativePathBuf, FileCache>>,
```

Comment:
One other consideration here: is it actually a requirement that we store the file contents in the cache? We're making the assumption that the file hasn't changed, so couldn't we just read these back from disk to report the diagnostics?

Reply:
I considered that. One challenge with this is that we perform transformations for jupyter notebooks. We would need to replicate that behavior.